File: MiniDOMOffice/CRM/Statistiche/StatisticheXPersona/Watchdogs/CPersonWatchCond.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.CustomerCalls;

namespace minidom
{
    public partial class CustomerCalls
    {

        /// <summary>
        /// Condizione di attenzione associata ad una persona
        /// </summary>
        [Serializable]
        public class CPersonWatchCond 
            : DMD.XML.DMDBaseXMLObject
        {
            private string m_SourceType;
            private int m_SourceID;
            private object m_Source;
            private string m_Descrizione;
            private int m_Gravita;
            private string m_Tag;
            private DateTime m_Data;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CPersonWatchCond()
            {
                m_SourceType = "";
                m_SourceID = 0;
                m_Source = null;
                m_Descrizione = "";
                m_Gravita = 0;
                m_Tag = "";
                m_Data = default;
            }

            /// <summary>
            /// Restituisce o imposta la data in cui si è verificata la condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public DateTime Data
            {
                get
                {
                    return m_Data;
                }

                set
                {
                    var oldValue = m_Data;
                    if (DMD.DateUtils.Compare(value, oldValue) == 0)
                        return;
                    m_Data = value;
                    DoChanged("Data", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il tipo dell'oggetto che ha generato la condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string SourceType
            {
                get
                {
                    if (m_Source is object)
                        return DMD.RunTime.vbTypeName(m_Source);
                    return m_SourceType;
                }

                set
                {
                    string oldValue = SourceType;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_SourceType = value;
                    DoChanged("SourceType", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'ID dell'oggetto che ha generato la condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int SourceID
            {
                get
                {
                    return DBUtils.GetID((Databases.IDBObjectBase)m_Source, m_SourceID);
                }

                set
                {
                    int oldValue = SourceID;
                    if (oldValue == value)
                        return;
                    m_SourceID = value;
                    m_Source = null;
                    DoChanged("SourceID", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'oggetto che ha generato la condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public object Source
            {
                get
                {
                    if (m_Source is null && !string.IsNullOrEmpty(m_SourceType))
                        m_Source = minidom.Sistema.ApplicationContext.GetItemByTypeAndId(m_SourceType, m_SourceID);
                    return m_Source;
                }

                set
                {
                    var oldValue = m_Source;
                    if (ReferenceEquals(oldValue, value))
                        return;
                    m_Source = value;
                    if (value is null)
                    {
                        m_SourceType = "";
                    }
                    else
                    {
                        m_SourceType = DMD.RunTime.vbTypeName(value);
                    }

                    m_SourceID = DBUtils.GetID(value, 0);
                    DoChanged("Source", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta una descrizione della condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Descrizione
            {
                get
                {
                    return m_Descrizione;
                }

                set
                {
                    string oldValue = m_Descrizione;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Descrizione = value;
                    DoChanged("Descrizione", value, oldValue);
                }
            }

            /// <summary>
            /// Restitusice o imposta dei parametri per la condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Tag
            {
                get
                {
                    return m_Tag;
                }

                set
                {
                    string oldValue = m_Tag;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Tag = value;
                    DoChanged("Tag", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la gravità (crescente) della condizione di attenzione
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int Gravita
            {
                get
                {
                    return m_Gravita;
                }

                set
                {
                    int oldValue = m_Gravita;
                    if (oldValue == value)
                        return;
                    m_Gravita = value;
                    DoChanged("Gravita", value, oldValue);
                }
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("SourceType", SourceType);
                writer.WriteAttribute("SourceID", SourceID);
                writer.WriteAttribute("Descrizione", m_Descrizione);
                writer.WriteAttribute("Tag", m_Tag);
                writer.WriteAttribute("Gravita", m_Gravita);
                writer.WriteAttribute("Data", m_Data);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "SourceType":
                        {
                            m_SourceType = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "SourceID":
                        {
                            m_SourceID = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Descrizione":
                        {
                            m_Descrizione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Tag":
                        {
                            m_Tag = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Gravita":
                        {
                            m_Gravita = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Data":
                        {
                            m_Data = (DateTime)DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

           
        }
    }
}

File: MiniDOMOffice/CRM/Statistiche/StatisticheXPersona/CPersonWatchCond.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases

Imports minidom.Anagrafica



Partial Public Class CustomerCalls

     

    Public Class CPersonWatchCond
        Inherits DBObjectBase

        Private m_SourceType As String
        Private m_SourceID As Integer
        Private m_Source As Object
        Private m_Descrizione As String
        Private m_Gravita As Integer
        Private m_Tag As String
        Private m_Data As Date

        Public Sub New()
            Me.m_SourceType = ""
            Me.m_SourceID = 0
            Me.m_Source = Nothing
            Me.m_Descrizione = ""
            Me.m_Gravita = 0
            Me.m_Tag = ""
            Me.m_Data = Nothing
        End Sub

        ''' <summary>
        ''' Restituisce o imposta la data in cui si è verificata la condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Data As Date
            Get
                Return Me.m_Data
            End Get
            Set(value As Date)
                Dim oldValue As Date = Me.m_Data
                If (DateUtils.Compare(value, oldValue) = 0) Then Exit Property
                Me.m_Data = value
                Me.DoChanged("Data", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il tipo dell'oggetto che ha generato la condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property SourceType As String
            Get
                If (Me.m_Source IsNot Nothing) Then Return TypeName(Me.m_Source)
                Return Me.m_SourceType
            End Get
            Set(value As String)
                Dim oldValue As String = Me.SourceType
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_SourceType = value
                Me.DoChanged("SourceType", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'ID dell'oggetto che ha generato la condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property SourceID As Integer
            Get
                Return GetID(Me.m_Source, Me.m_SourceID)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.SourceID
                If (oldValue = value) Then Exit Property
                Me.m_SourceID = value
                Me.m_Source = Nothing
                Me.DoChanged("SourceID", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'oggetto che ha generato la condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Source As Object
            Get
                If (Me.m_Source Is Nothing AndAlso Me.m_SourceType <> "") Then Me.m_Source = Sistema.Types.GetItemByTypeAndId(Me.m_SourceType, Me.m_SourceID)
                Return Me.m_Source
            End Get
            Set(value As Object)
                Dim oldValue As Object = Me.m_Source
                If (oldValue Is value) Then Exit Property
                Me.m_Source = value
                If (value Is Nothing) Then
                    Me.m_SourceType = ""
                Else
                    Me.m_SourceType = TypeName(value)
                End If
                Me.m_SourceID = GetID(value)
                Me.DoChanged("Source", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta una descrizione della condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Descrizione As String
            Get
                Return Me.m_Descrizione
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_Descrizione
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_Descrizione = value
                Me.DoChanged("Descrizione", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restitusice o imposta dei parametri per la condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Tag As String
            Get
                Return Me.m_Tag
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_Tag
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_Tag = value
                Me.DoChanged("Tag", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la gravità (crescente) della condizione di attenzione
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Gravita As Integer
            Get
                Return Me.m_Gravita
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_Gravita
                If (oldValue = value) Then Exit Property
                Me.m_Gravita = value
                Me.DoChanged("Gravita", value, oldValue)
            End Set
        End Property

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("SourceType", Me.SourceType)
            writer.WriteAttribute("SourceID", Me.SourceID)
            writer.WriteAttribute("Descrizione", Me.m_Descrizione)
            writer.WriteAttribute("Tag", Me.m_Tag)
            writer.WriteAttribute("Gravita", Me.m_Gravita)
            writer.WriteAttribute("Data", Me.m_Data)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "SourceType" : Me.m_SourceType = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "SourceID" : Me.m_SourceID = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Descrizione" : Me.m_Descrizione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Tag" : Me.m_Tag = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Gravita" : Me.m_Gravita = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Data" : Me.m_Data = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub


        Protected Overrides Function GetConnection() As CDBConnection
            Return Nothing
        End Function

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overrides Function GetTableName() As String
            Return ""
        End Function



    End Class



End Class